src/launch/nss-cache.c

```diff
@@ -114,45 +114,50 @@ static int nss_cache_add(CRBTree *tree_by_name,
                          uint32_t id) {
         CRBNode **slot_by_name, **slot_by_id, *parent_by_name, *parent_by_id;
-        NSSCacheNode *node_by_name, *node_by_id;
+        NSSCacheNode *node_by_name = NULL, *node_by_id = NULL;
         int r;
 
-        slot_by_name = c_rbtree_find_slot(tree_by_name,
-                                          nss_cache_node_compare_name,
-                                          name,
-                                          &parent_by_name);
-        slot_by_id = c_rbtree_find_slot(tree_by_id,
-                                        nss_cache_node_compare_id,
-                                        (void *)(unsigned long)id,
-                                        &parent_by_id);
-
-        if (slot_by_name)
-                node_by_name = NULL;
-        else
-                node_by_name = c_rbnode_entry(parent_by_name, NSSCacheNode, rb_by_name);
+        if (name) {
+                slot_by_name = c_rbtree_find_slot(tree_by_name,
+                                                  nss_cache_node_compare_name,
+                                                  name,
+                                                  &parent_by_name);
 
-        if (slot_by_id)
-                node_by_id = NULL;
-        else
-                node_by_id = c_rbnode_entry(parent_by_id, NSSCacheNode, rb_by_id);
+                if (!slot_by_name)
+                        node_by_name = c_rbnode_entry(parent_by_name, NSSCacheNode, rb_by_name);
+        }
+
+        if (id != -1) {
+                slot_by_id = c_rbtree_find_slot(tree_by_id,
+                                                nss_cache_node_compare_id,
+                                                (void *)(unsigned long)id,
+                                                &parent_by_id);
+                if (!slot_by_id)
+                        node_by_id = c_rbnode_entry(parent_by_id, NSSCacheNode, rb_by_id);
+        }
 
         /*
          * Either the entry is linked by-name *AND* by-id (and it is the *SAME*
-         * entry), or it is not linked at all. However, if only one of both is
-         * known, or if they point to different entries, we will *NOT* cache
-         * the new data. This situation implies a UID/GID conflict and there is
-         * no sane way to handle this, other than bypassing the cache. Worst
-         * case, you will end up calling into NSS all the time.
+         * entry), it is linked by one and the other is invalid, or it is not
+         * linked at all. However, if only one of both is known and valid, or
+         * if they point to different entries, we will *NOT* cache the new data.
+         * This situation implies a UID/GID conflict and there is no sane way to
+         * handle this, other than bypassing the cache. Worst case, you will
+         * end up calling into NSS all the time.
          */
-        if (node_by_name != node_by_id) {
+        if ((node_by_id && node_by_id->name && !node_by_name) ||
```

Comment:
This is wrong. `name` is embedded, it cannot be NULL. You *MUST* check for `node->pw.pw_name`, but you can't since you don't know whether this is a user or group. Why not rather check `c_rbnode_is_linked(&node->rb_by_name)`?

Reply:
Fixed.